core/src/main/java/io/questdb/cutlass/line/tcp/LineTcpConnectionContext.java

```diff
@@ -0,0 +1,148 @@
+package io.questdb.cutlass.line.tcp;
+
+import io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler.LineTcpMeasurementEvent;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.network.IOContext;
+import io.questdb.network.IODispatcher;
+import io.questdb.network.IOOperation;
+import io.questdb.network.NetworkFacade;
+import io.questdb.std.Mutable;
+import io.questdb.std.Unsafe;
+import io.questdb.std.str.DirectByteCharSequence;
+
+class LineTcpConnectionContext implements IOContext, Mutable {
+    private static final Log LOG = LogFactory.getLog(LineTcpConnectionContext.class);
+    private final NetworkFacade nf;
+    private final LineTcpMeasurementScheduler scheduler;
+    private long fd;
+    private IODispatcher<LineTcpConnectionContext> dispatcher;
+    private long recvBufStart;
+    private long recvBufEnd;
+    private long recvBufPos;
+    private boolean peerDisconnected;
+    private final DirectByteCharSequence byteCharSequence = new DirectByteCharSequence();
+
+    LineTcpConnectionContext(LineTcpReceiverConfiguration configuration, LineTcpMeasurementScheduler scheduler) {
+        nf = configuration.getNetworkFacade();
+        this.scheduler = scheduler;
+        recvBufStart = Unsafe.malloc(configuration.getNetMsgBufferSize());
+        recvBufEnd = recvBufStart + configuration.getNetMsgBufferSize();
+    }
+
+    // returns true if busy
+    boolean handleIO() {
+        try {
+            LineTcpMeasurementEvent event = scheduler.getNewEvent();
```

Comment:
There is a trade off here, we can either always commit the queue state quickly (within one call to handelO()) or propagate our waiting state back to the TCP peer producer by not reading from the TCP socket. Since the Sequence  implementation is for a single producer, holding onto the sequence cursor for long has no bad effects.